src/hivpy/sexual_behaviour.py

```diff
@@ -130,3 +144,143 @@ def update_sex_groups(self, population: pd.DataFrame):
+            # This might be more efficient, but is also a bit obscure
+            # = ltp_risk_factor if ltp is true, and 1 if ltp is false.
+            # pop_data["rred_ltp"] = pop_data["ltp"]*self.ltp_risk_factor+(1-pop_data["ltp"])
+
+    def init_rred_personal(self, population, n_pop):
+        p_rred_p = self.sb_data.p_rred_p_dist.sample()
+        population["rred_personal"] = np.ones(n_pop)
+        r = rng.uniform(size=n_pop)
+        mask = r < p_rred_p
+        population.loc[mask, "rred_personal"] = 1e-5
+
+    def init_rred_adc(self, population):
+        population["rred_adc"] = 1.0
+        self.update_rred_adc(population)
+
+    def update_rred_adc(self, population):
+        """Updates risk reduction for AIDS defining condition"""
+        # We don't need the rred_adc==1 condition (since they are all set to rred_adc = 1 to start)
+        # It prevents needless assignments but requires checking more conditions
+        # Not sure which is more efficient or if it matters.
+        indices = selector(population, rred_adc=(operator.eq, 1), HIV_status=(operator.eq, True))
+        population.loc[indices, "rred_adc"] = 0.2
+
+    def init_rred_population(self, pop_data):
+        """Initialise general population risk reduction w.r.t. condomless sex with new partners"""
+        self.rred_population = 1
+
+    def update_rred_population(self, pop_data, date):
+        date1995 = datetime.date(1995, 1, 1)
+        date2000 = datetime.date(2000, 1, 1)
+        date2010 = datetime.date(2010, 1, 1)
+        date2021 = datetime.date(2021, 1, 1)
+        yearly_change_90s = self.yearly_risk_change["1990s"]
+        yearly_change_10s = self.yearly_risk_change["2010s"]
+        if(date1995 < date <= date2000):
+            # there ought to be a better way to get the fractional number of years
+            dt = (date - date1995) / datetime.timedelta(days=365.25)
+            self.rred_population = yearly_change_90s**dt
+        elif(date2000 < date < date2010):
+            self.rred_population = yearly_change_90s**5
+        elif(date2010 < date < date2021):
+            dt = (date - date2010) / datetime.timedelta(days=365.25)
+            self.rred_population = yearly_change_90s**5 * yearly_change_10s**dt
+        elif(date2021 < date):
+            self.rred_population = yearly_change_90s**5 * yearly_change_10s**11
+
+    def init_rred_diagnosis(self, population):
+        population["rred_diagnosis"] = 1
+
+    def update_rred_diagnosis(self, population, date):
+        HIV_idx_new = selector(population, HIV_status=(operator.eq, True),
+                               HIV_Diagnosis_Date=(operator.ge, date-self.rred_diagnosis_period))
+        population.loc[HIV_idx_new, "rred_diagnosis"] = self.rred_diagnosis
+        HIV_idx_old = selector(population, HIV_status=(operator.eq, True),
+                               HIV_Diagnosis_Date=(operator.lt, date-self.rred_diagnosis_period))
```

Comment:
Just thinking here. I wonder if it's worth recording these variables ("recent diagnosis", "old diagnosis") on the data frame too. But then we'd need to keep the values updated. It may speed this up a bit but probably not significantly. Likely not worth it if they're not used elsewhere (with the same period).

Reply:
I'm don't know whether that it would speed it up if the categories couldn't be used elsewhere; on the one hand you would only have to update people with a "recent diagnosis" because old diagnoses will stay old; then in this function we only need to update the `rred_diagnosis` value for people who have changed from recent to old in this time step, or from no diagnosis to recent. So to some extent this would then depend on how efficiently we compare the population's current state to previous state. I would be tempted to keep a note of this but stick a pin in it for now?